### PR TITLE
#1131 Make toolbar size customizable

### PR DESCRIPTION
**Pull request by nabilIbm**
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Review by tomlyn — CHANGES_REQUESTED** [review]
Code looks good unfortunately I tried the code and something looks wrong. In the test harness using 
structuretable_paramDef > Table > Configure Custom Table Buttons
Here's how it looks with your changes: 
image

but without your changes it looks like this:
image

File: canvas_modules/common-canvas/__tests__/toolbar/toolbar-test.js [hunk 1]
```diff
@@ -107,15 +107,60 @@ describe("Toolbar renders correctly", () => {
 			.simulate("click");
 		expect(toolbarActionHandler.calledOnce).to.equal(false);
 	});
+
+	it("should render a Toolbar with medium size buttons", () => {
```
Comment (caritaou) [inline canvas_modules/common-canvas/__tests__/toolbar/toolbar-test.js:111]: Can you add tests to make sure `hasIconOnly` get applied to the buttons correctly for those without a label?

**Review by caritaou — CHANGES_REQUESTED** [review]
image

The styles are wrong, it does not look identical to what's currently in prod. Looking at the DOM, `hasIconOnly` did not get applied to these buttons

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss [hunk 2]
```diff
@@ -35,16 +32,26 @@ $toolbar-divider-width: 1px;
 
 	/* Prevent elements from being dragged onto canvas */
 	-webkit-user-drag: none;
+	.toolbar-left-bar
+	.toolbar-item.default.icon_button_1-action button,
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/toolbar/toolbar.scss:36]: `icon_button_*-action` classnames are specific to the test added in structuretable_paramDef.json for table buttons. These classnames should not be referenced for styling. I also don't see the styles applied in the DOM, are these needed?

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss [hunk 3]
```diff
@@ -68,12 +74,15 @@ $toolbar-divider-width: 1px;
 	&.ghost {
 		& button {
 			background-color: $ui-01;
+			padding: 0;
+			.toolbar-icon-label.before {
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/toolbar/toolbar.scss:78]: The styling here does not look correct, the label and icon should be vertically centered. We need to also use the Carbon space tokens instead of specifying an arbitrary value like 5 or 6

image

**Review by caritaou — CHANGES_REQUESTED** [review]
With your new changes, you should be able to remove https://github.com/elyra-ai/canvas/blob/main/canvas_modules/common-canvas/src/common-properties/components/table-buttons/table-buttons.scss and everything should still look the same since the styles should now be handled by the Toolbar and not the table-button from within CommonProperties.

image

**Comment by nabilIbm** [comment]
Hi @caritaou This is the view after if I delete `table-buttons.scss `
image

**Comment by caritaou** [comment]
Hi @nabilIbm, this issue is to remove the styles that table-buttons.scss is specifying for the toolbar and have the toolbar handle the styles for the different toolbar sizes. So you will need to fix the styles for the toolbar after removing the table-buttons.scss. You can take a look at the styles and see if you can port them over

File: canvas_modules/common-canvas/__tests__/toolbar/toolbar-test.js [hunk 4]
```diff
@@ -107,15 +107,90 @@ describe("Toolbar renders correctly", () => {
...
+			leftBar: [
+				{ action: "run", label: "Label btn", enable: true },
+			]
+		};
+		const canvasToolbar = createToolbar(toolbarConfig);
+		// Select the toolbar only icons buttons
+		const defaultButtons = canvasToolbar.find(".toolbar-item.default button");
+
+		expect(defaultButtons).to.have.length(1);
+		expect(defaultButtons.find(".bx--btn--icon-only")).to.have.length(1);
```
Comment (caritaou) [inline canvas_modules/common-canvas/__tests__/toolbar/toolbar-test.js:182]: This test says testing icon with label yet it is expecting to find `".bx--btn--icon-only"`

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss [hunk 5]
```diff
@@ -35,6 +32,9 @@ $toolbar-divider-width: 1px;
 
 	/* Prevent elements from being dragged onto canvas */
 	-webkit-user-drag: none;
+	.toolbar-left-bar .toolbar-item-content.default {
+		padding: 0;
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/toolbar/toolbar.scss:36]: Why is the padding set to 0? This is affecting the medium size toolbar

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss [hunk 6]
```diff
@@ -43,6 +43,7 @@ $toolbar-divider-width: 1px;
 	padding-right: 42px;
 	overflow: hidden;
 	align-items: center;
+	padding-right: 0;
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/toolbar/toolbar.scss:46]: This is causing issues with the overflow icon

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss [hunk 7]
```diff
@@ -68,12 +68,17 @@ $toolbar-divider-width: 1px;
 	&.ghost {
 		& button {
 			background-color: $ui-01;
+			padding: 0;
+			.toolbar-item-content {
+				padding: 11px;
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/toolbar/toolbar.scss:73]: I dont see this padding being applied. We should also stay away from hardcoded values, try to use Carbon tokens instead

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss [hunk 8]
```diff
@@ -105,56 +110,45 @@ $toolbar-divider-width: 1px;
...
 	}
 }
-
 .toolbar-overflow-menu-item button {
 	width: 100%;
 	padding-top: 2px;
 	padding-bottom: 2px;
 	min-height: 20px;
 }
-
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/toolbar/toolbar.scss:137]: Can you undo these changes? Its more clear to have the newlines while reading this file

**Review by caritaou — CHANGES_REQUESTED** [review]
image
There are many issues with the default medium sized toolbar now. The overflow icon size is incorrect and does not appear at times when resizing the canvas

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss (hunk 8 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/toolbar/toolbar.scss:137]: The reason to chose these unique classes is, to separate my style from the icons within the toolbar in canvas and other toolbar icons in table buttons since they both share pretty much the same classes except the ones I chose.

```
.togglePalette-action,
	.undo-action,.redo-action,
	.cut-action,.copy-action,
	.paste-action,
	.createAutoComment-action,
	.deleteSelectedObjects-action{
		padding: 3px;
	}
```
Looks like these extra new line has not been fixed...I'll see what else I can do for it

Comment (nabilIbm) [reply canvas_modules/common-canvas/src/toolbar/toolbar.scss:137]: > image
> 
The reason to chose these unique classes is, to separate my style from the icons within the toolbar in canvas and other toolbar icons in table buttons since they both share pretty much the same classes except the ones I chose.

.togglePalette-action,
	.undo-action,.redo-action,
	.cut-action,.copy-action,
	.paste-action,
	.createAutoComment-action,
	.deleteSelectedObjects-action{
		padding: 3px;
	}
Looks like these extra new line has not been fixed...I'll see what else I can do for it

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss [hunk 9]
```diff
@@ -40,9 +37,19 @@ $toolbar-divider-width: 1px;
...
+	.toolbar-icon-label.before {
+		padding: $spacing-04;
+	}
+	.togglePalette-action,
+	.undo-action,.redo-action,
+	.cut-action,.copy-action,
+	.paste-action,
+	.createAutoComment-action,
+	.deleteSelectedObjects-action{
+		padding: 3px;
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/toolbar/toolbar.scss:51]: You cant specify the styles like this for the specific toolbar items. These are some of the default ones that canvas provides, but applications can customize it with their own items. Any changes done to toolbar items need to be generic

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss (hunk 8 above)
Comment (caritaou) [reply canvas_modules/common-canvas/src/toolbar/toolbar.scss:137]: >The reason to chose these unique classes is, to separate my style from the icons within the toolbar in canvas and other toolbar icons in table buttons

You can add additional class name to differentiate when toolbar is inside a table or not

**Review by caritaou — CHANGES_REQUESTED** [review]

File: canvas_modules/common-canvas/src/common-properties/components/table-buttons/table-buttons.jsx [hunk 10]
```diff
@@ -122,7 +123,7 @@ TableButtons.propTypes = {
 	propertyId: PropTypes.object.isRequired,
 	customButtons: PropTypes.array.isRequired,
 	tableState: PropTypes.string,
-	customButtonsState: PropTypes.object // set in by redux
+	customButtonsState: PropTypes.object, // set in by redux
```
[the remark this answers is not in the document]
Comment (caritaou) [reply canvas_modules/common-canvas/src/common-properties/components/table-buttons/table-buttons.jsx:126]: This was not addressed, marking as unresolved

File: canvas_modules/common-canvas/src/toolbar/toolbar-action-item.jsx [hunk 11]
```diff
@@ -175,20 +175,22 @@ class ToolbarActionItem extends React.Component {
 				{textContent}
 			</div>
 		);
-
 		buttonContent = this.wrapInTooltip(buttonContent);
 
+		let btnHasIcon = false;
```
[the remark this answers is not in the document]
Comment (caritaou) [reply canvas_modules/common-canvas/src/toolbar/toolbar-action-item.jsx:180]: This was not addressed, marking as unresolved

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss (hunk 8 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/toolbar/toolbar.scss:137]: image

File: canvas_modules/common-canvas/src/common-properties/components/table-buttons/table-buttons.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/common-properties/components/table-buttons/table-buttons.jsx]: image

**Review by caritaou — CHANGES_REQUESTED** [review]
overflow icon is still incorrect. There is also an issue with the notification icon now, the number is off centered
image

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss (hunk 8 above)
Comment (caritaou) [reply canvas_modules/common-canvas/src/toolbar/toolbar.scss:137]: This will change everything. You can create a new class name in the code specifically for toolbar in tables, that way the styles will only apply to that instead of the canvas toolbar too

File: canvas_modules/common-canvas/src/toolbar/toolbar.jsx [hunk 12]
```diff
@@ -279,10 +281,10 @@ class Toolbar extends React.Component {
 
 		const leftItems = this.generateToolbarItems(this.leftBar, false, true);
 		const rightItems = this.generateToolbarItems(this.rightBar, false, false);
-
+		const toolbarSizeClass = this.props.size === "sm" ? "toolbar-div toolbar-size-small" : "toolbar-div";
 		const canvasToolbar = (
 			<ReactResizeDetector handleWidth onResize={this.onToolbarResize}>
-				<div className="toolbar-div" instanceid={this.props.instanceId}>
+				<div className={toolbarSizeClass} toolbarSizeClass instanceid={this.props.instanceId}>
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/toolbar/toolbar.jsx:287]: There is an extra `toolbarSizeClass` here

**Review by tomlyn — APPROVED** [review]

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss [hunk 13]
```diff
@@ -35,6 +35,70 @@ $toolbar-divider-width: 1px;
...
+		.toolbar-left-bar {
+			align-items: center;
+			text-align: center;
+			padding: 0px;
+		}
+		.toolbar-item {
+			&.default,
+			&.ghost {
+				& button {
+					padding: 3px;
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/toolbar/toolbar.scss:48]: Why is this only for default and ghost types? 
The button is wrong. Ypu can see the height extends past the toolbar 
image
image

**Review by caritaou — CHANGES_REQUESTED** [review]
There are alot of css changes applied to the small toolbar that is unneceaasry. The simplest solution is to copy over the styles deleted from table-buttons.scss for the left toolbar.

Revisiting the table-buttonns.scss, we actually still need to hide the right side toolbar from inside common properties tables
```
.toolbar-right-bar {
	display: none;
}
```

File: canvas_modules/common-canvas/src/common-properties/components/table-buttons/table-buttons.scss [hunk 14]
```diff
@@ -1,49 +0,0 @@
...
-				align-items: center;
-			}
-
-			.toolbar-overflow-item button {
-				height: $spacing-07;
-			}
-
-			.toolbar-popover-list {
-				top: $spacing-07;
-				right: 0;
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/common-properties/components/table-buttons/table-buttons.scss:41]: Hey Nabil, we will still need this style for the overflow icon inside tables, so the menu will appear to the left instead of the right
image

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss [hunk 15]
```diff
@@ -35,6 +35,70 @@ $toolbar-divider-width: 1px;
...
+				& svg {
+					height: $toolbar-icon-size;
+					width: $toolbar-icon-size;
+					/* Prevent elements from being dragged onto canvas */
+					-webkit-user-drag: none;
+				}
+			}
+		}
+	}
 }
```
Comment (tomlyn) [inline canvas_modules/common-canvas/src/toolbar/toolbar.scss:102]: Regarding Carita's comments above. I think what you ought to do is take the SCSS that is in table-buttons.scss as a starting point for this SCSS.  You can insert that SCSS inside this section: 
```
&.toolbar-size-small {
}
```
This will isolate those styles from the SCSS for medium size buttons. Then do some testing to see if things are working when the "small" size toolbar is displayed in common-properties.

**Review by tomlyn — CHANGES_REQUESTED** [review]

File: canvas_modules/common-canvas/src/common-properties/components/table-buttons/table-buttons.scss (hunk 14 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/common-properties/components/table-buttons/table-buttons.scss:41]: image

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss (hunk 13 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/toolbar/toolbar.scss]: image

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss [hunk 16]
```diff
@@ -35,6 +35,70 @@ $toolbar-divider-width: 1px;
...
+			display: inline-block;
+			& button {
+				padding:0;
+				height: $toolbar-button-height;
+				min-height: 30px;
+				background-color: $ui-01;
+				border-right: $toolbar-divider-width solid $ui-03;
+			}
+			& button:hover {
+				background-color: $hover-field;
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/toolbar/toolbar.scss:63]: Hi Carita, this is the hover effect for the button in `.toolbar-overflow-item` it's already in the default  I just added it to the small size cuz,  if I disregard it here it the hover won't work in a small size

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss [hunk 17]
```diff
@@ -35,6 +35,70 @@ $toolbar-divider-width: 1px;
...
+			}
+			& button:disabled:hover {
+				background-color: $ui-01;
+			}
+		}
+		.toolbar-overflow-menu-item button {
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/toolbar/toolbar.scss:69]: image

File: canvas_modules/common-canvas/src/toolbar/toolbar-overflow-item.jsx [hunk 18]
```diff
@@ -53,6 +53,7 @@ class ToolbarOverflowItem extends React.Component {
 						onClick={this.toggleExtendedMenu}
 						onFocus={this.props.onFocus}
 						aria-label={this.props.label}
+						size={this.props.size}
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/toolbar/toolbar-overflow-item.jsx:56]: You need to set `hasIconOnly` in the props here. You shouldnt need to have so many style overrides just for this button.

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss (hunk 16 above)
[the remark this answers is not in the document]
Comment (caritaou) [reply canvas_modules/common-canvas/src/toolbar/toolbar.scss:64]: This doens't sound right, why does changing the size of the buttons affect the colors at all. You set the backgound-color on line 60, which in turn affects the hover color.  Can you try removing all the color overrides you added for small buttons and see what the behavior is?

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss [hunk 19]
```diff
@@ -22,6 +22,9 @@ $toolbar-button-height: 40px;
 $toolbar-button-width: 40px;
 $toolbar-divider-width: 1px;
 
+.properties-custom-table-buttons .toolbar-right-bar {
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/toolbar/toolbar.scss:25]: `properties-custom-table-buttons` is a class name inside common properties. It should not be part of the toolbar styles. 

> Revisiting the table-buttons.scss, we actually still need to hide the right side toolbar from inside common properties tables

This change is specific to table-buttons and should not be a part of the toolbar. Having this in the toolbar styles is saying that any small sized toolbars will automatically hide the right side items. Can you move this to the table-buttons? Also lets keep the css selector to be more define 

```
.properties-custom-table-buttons {
	.toolbar-div {
		.toolbar-right-bar {
			display: none;
		}
	}
}
```

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss [hunk 20]
```diff
@@ -35,6 +38,61 @@ $toolbar-divider-width: 1px;
 
 	/* Prevent elements from being dragged onto canvas */
 	-webkit-user-drag: none;
+	&.toolbar-size-small {
+		.toolbar-left-bar {
+			padding: 0;
+			.toolbar-popover-list {
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/toolbar/toolbar.scss:44]: This change is also specific to table-buttons as well. We only want the menu of toolbars inside table-buttons to appear on the left. This should not affect any toolbar of size small

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss [hunk 21]
```diff
@@ -35,6 +38,61 @@ $toolbar-divider-width: 1px;
...
+			}
+			& button:disabled:hover {
+				background-color: $ui-01;
+			}
+		}
+		.toolbar-overflow-menu-item button {
+			width: 100%;
+			padding-top: 2px;
+			padding-bottom: 2px;
+			min-height: 20px;
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/toolbar/toolbar.scss:74]: There are many hardcoded `px` values that should not be necessary if you are able to leverage the Carbon styles for buttons. Can you revisit these and see what you can do? Please see this comment https://github.com/elyra-ai/canvas/pull/1143/files#r1009987416

**Review by caritaou — CHANGES_REQUESTED** [review]

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss (hunk 19 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/toolbar/toolbar.scss]: same comment below

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss (hunk 20 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/toolbar/toolbar.scss]: moved into the `table-buttons.scss `
image

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss (hunk 21 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/toolbar/toolbar.scss]: I did remove some hardcoded...but I added a few styles to make it identical to the cloud, also you commented about the backgroundColor within `&.toolbar-size-small`, the thing is if i remove it i lose  overflowicon color oh hover 
image

**Review by caritaou — CHANGES_REQUESTED** [review]
For the toolbar styles, use this:
```
        &.toolbar-size-small {
                height: $spacing-07;
                .toolbar-left-bar {
                        padding: 0;
                }
                .toolbar-item button,
                .toolbar-overflow-item > button {
                        padding: 0;
                        height: $spacing-07;
                }
                .toolbar-item-content {
                        padding: $spacing-02 $spacing-03;
                        height: $spacing-07;
                        align-items: center;
                }
                .toolbar-popover-list button {
                        padding: 0;
                }
        }
```

File: canvas_modules/common-canvas/src/toolbar/toolbar.scss (hunk 21 above)
Comment (caritaou) [reply canvas_modules/common-canvas/src/toolbar/toolbar.scss]: All styles need to conform to Carbon standards, you can take a look at what styles are applied for these icons for future reference https://v7-react.carbondesignsystem.com/?path=/story/components-button--playground

**Review by caritaou — APPROVED** [review]

**Review by tomlyn — APPROVED** [review]
